pip-lock/main.ts

```diff
@@ -0,0 +1,392 @@
+
+// Extracts the package version from the target string, which is of the form:
+// Flask-2.2.2-py3-none-any.whl
+// This is made complicated due to the fact that package names may not match
+// the provided packageName, and can contain -, _, ., etc. Versions may not
+// adhere to pep8 standards also, so that must be accounted for here.
+// This method will ultimately return the package/version tuple.
+function parsePackageVersion(packageName: string, target: string): string[] {
+    if(target.length < (packageName.length+2))
+        throw new Error(`invalid target ${target} does not contain the package name ${packageName}`);
+
+    // We will parse the real package name out of the target string
+    // if it isn't the same as the package name provided. This happens with some packages -
+    // consider flask shows up as Flask here. At present, our heuristic is that the beginning of the
+    // next field (which is the package version) must start with the pattern `-[0-9]`; there are no
+    // other guarantees, so that is what we start by looking for.
+    let fixedName = packageName;
+    if(!target.startsWith(packageName)) {
+        const tmp: string[] = [];
+        for(let i = 0; i < target.length; i++) {
+            if(target[i] === '-' && i < (target.length - 1)) {
+                const check = target[i+1];
+                // If value is ascii '0'-'9'
+                if(check.charCodeAt(0) <= 57 && check.charCodeAt(0) >= 48)
+                    break;
+            }
+
+            tmp.push(target[i]);
+        }
+
+        fixedName = tmp.join('');
+    }
+    // Trim off the first portion containing the parsed name
+    const rem = target.substring(fixedName.length+1);
+    const [versionFirstPart, ...rest] = rem.split('-');
+
+    // Now we will validate the version portion of the name.
+    // Some versions contain weird artifacts - like binary extensions - at the end. We
+    // will make a best effort at cleaning this up.
+    const verList: string[] = [];
+    for(let i = 0; i < versionFirstPart.length; i++) {
+        if (versionFirstPart[i] === '.' && i < (versionFirstPart.length - 1)) {
+            // Check to see if the next part of the value is non-numeric. This may not mean that we are
+            // dealing with something we need to get rid of (as versions _could_ have rc or alpha versions or similar),
+            // but we will now need to check.
+            if (versionFirstPart[i + 1].charCodeAt(0) > 57 || versionFirstPart[i + 1].charCodeAt(0) < 48) {
+                const nextValue = versionFirstPart.substring(i);
+                // Now we need to check to see if this is a raw binary file or not - some "version" strings
+                // are simply just a tarball or wheel file.
+                if (binaryFormats.has(nextValue))
+                    break;
+            }
+        }
+
+        verList.push(versionFirstPart[i]);
+    }
+
+    const verFixed = verList.join('');
```

Comment:
Can all of this be replaced by `const verFixed = versionFirstPart.replace(/-(?:\.tar\.gz|\.zip|\.whl|\.pyd)/, '');`?

Reply:
We are just parsing things from messy output, not trying to clean them up, so replace is probably not the right choice here.